runtime/service/src/main/java/org/apache/polaris/service/events/listeners/inmemory/InMemoryEventListener.java

```diff
@@ -0,0 +1,141 @@
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import com.google.common.annotations.VisibleForTesting;
+import io.smallrye.common.annotation.Identifier;
+import io.smallrye.mutiny.infrastructure.Infrastructure;
+import io.smallrye.mutiny.operators.multi.processors.UnicastProcessor;
+import jakarta.annotation.Nullable;
+import jakarta.annotation.PreDestroy;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import jakarta.ws.rs.container.ContainerRequestContext;
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.core.SecurityContext;
+import java.time.Clock;
+import java.time.Duration;
+import java.util.List;
+import java.util.Objects;
+import org.apache.polaris.core.PolarisCallContext;
+import org.apache.polaris.core.context.CallContext;
+import org.apache.polaris.core.context.RealmContext;
+import org.apache.polaris.core.entity.PolarisEvent;
+import org.apache.polaris.core.persistence.MetaStoreManagerFactory;
+import org.apache.polaris.service.events.listeners.InMemoryBufferEventListenerConfiguration;
+import org.apache.polaris.service.events.listeners.PolarisPersistenceEventListener;
+import org.eclipse.microprofile.faulttolerance.Fallback;
+import org.eclipse.microprofile.faulttolerance.Retry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ApplicationScoped
+@Identifier("persistence-in-memory")
+public class InMemoryEventListener extends PolarisPersistenceEventListener {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(InMemoryEventListener.class);
+
+  @Inject CallContext callContext;
+  @Inject Clock clock;
+  @Inject MetaStoreManagerFactory metaStoreManagerFactory;
+  @Inject InMemoryBufferEventListenerConfiguration configuration;
+
+  @Context SecurityContext securityContext;
+  @Context ContainerRequestContext requestContext;
+
+  @VisibleForTesting
+  final LoadingCache<String, UnicastProcessor<PolarisEvent>> processors =
+      Caffeine.newBuilder()
+          .expireAfterAccess(Duration.ofHours(1))
+          .evictionListener(
+              (String realmId, UnicastProcessor<?> processor, RemovalCause cause) ->
+                  processor.onComplete())
+          .build(this::createProcessor);
+
+  @Override
+  protected void processEvent(PolarisEvent event) {
+    var realmId = callContext.getRealmContext().getRealmIdentifier();
+    processEvent(realmId, event);
+  }
+
+  protected void processEvent(String realmId, PolarisEvent event) {
+    var processor = Objects.requireNonNull(processors.get(realmId));
+    processor.onNext(event);
+  }
+
+  @Override
+  protected ContextSpecificInformation getContextSpecificInformation() {
+    var principal = securityContext.getUserPrincipal();
+    var principalName = principal == null ? null : principal.getName();
+    return new ContextSpecificInformation(clock.millis(), principalName);
```

Comment:
may be this is not in the scope of the pr, but do we know why just principal and not the activated roles ?